### PR TITLE
archive: remove path folder itself when remove paramater is true

### PR DESCRIPTION
**Pull request by aminvakil**
##### SUMMARY
Fixes https://github.com/ansible-collections/community.general/issues/1041

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
archive

### CODE REVIEW

File: changelogs/fragments/1083-archive-remove-path-folder.yml
Comment (aminvakil) [inline changelogs/fragments/1083-archive-remove-path-folder.yml:2]: Maybe someone has been relying on this behaviour and this is a `breaking_change` for them, so I don't know if this should change and wait for the next major release of `community.general` or not.

File: plugins/modules/files/archive.py
Comment (felixfontein) [inline plugins/modules/files/archive.py:459]: This also removes other files which were not added to the archive (`exclude_path` option).

How about the following:
1. Sort b_expanded_paths reversely, and iterate over it that way. That should make sure that child directories come before parent directories.
2. Also use `os.scandir(b_path)` to check whether the directory is empty (resp. contains more than `.` and `..`). If it does, skip the directory.
3. Use `os.rmdir`, which only removes the directory if they are empty.

File: changelogs/fragments/1083-archive-remove-path-folder.yml
Comment (felixfontein) [reply changelogs/fragments/1083-archive-remove-path-folder.yml:2]: Since the documentation mentions this is supposed to remove trees as well, I think it is OK to do this as a bugfix. If the documentation would not mention that, it would definitely be a breaking change (or would require an option that needs to be turned on explicitly).

File: plugins/modules/files/archive.py
Comment (aminvakil) [reply plugins/modules/files/archive.py:459]: This led me to another thing which I'm not sure if this is a bug or not.
If you specify the folder in path, you cannot exclude a file in the folder.
```
# does not exclude
path: tmpdir
exclude_path: tmpdir/a.txt
```
```
# excludes
path: tmpdir/*
exclude_path: tmpdir/a.txt
```
So if this behaviour is expected and ok, then the current way of removing path when `remove: yes` is ok.

Comment (aminvakil) [reply plugins/modules/files/archive.py:459]: This has been addressed on https://github.com/ansible-collections/community.general/issues/593 as not a bug.
What do you suggest to do now?

Comment (felixfontein) [reply plugins/modules/files/archive.py:459]: Hmm. I guess in that case it is ok.

**Comment by felixfontein** [comment]
I'm wondering whether we should delay this fix to 2.0.0, instead of adding it to 1.3.0, and marking it also with `breaking_changes`. If suddenly something is deleted that has not been deleted in the past, that could be very surprising to users. On the other hand, the current behavior is obviously wrong and different from the documented one.

@goniomdq since you reported this issue (and apparently use the `remove` option :) ), what do you think?

**Comment by goniomdq** [comment]
Sorry, but I'm not sure I understand why the delay. 
Are you saying that the fix breaks something else?
Also, when is 2.0.0 due to be released and why leaving it to then?
Thank you.

**Comment by felixfontein** [comment]
Well, this is about deleting things, and with the change the module deletes more things than before. It could be that some people got used to the module not deleting certain things, and for them this could be dangerous.

community.general 2.0.0 is due at the beginning of next year, and will appear in Ansible 2.11.0 (also due then).

**Comment by aminvakil** [comment]
Per discussion in [community meeting](https://github.com/ansible/community/issues/539#issuecomment-715496088) it would be better to wait until next major release which it's ok to have breaking changes, so this would get merged in 2.0.0.

**Comment by felixfontein** [comment]
Closing+Reopening to re-trigger tests

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@aminvakil thanks for fixing this!

I marked it as a breaking change so it won't get accidentally backported, so the first release it shows up in will be 2.0.0.

**Comment by aminvakil** [comment]
@felixfontein Great, thanks for reviewing!

**Comment by aminvakil** [comment]
@felixfontein 

Oops! Shouldn't have I changed the `changelog`? From `bugfixes` to `breaking_change`?

**Comment by felixfontein** [comment]
@aminvakil hmm, yeah, might be a good idea. Can you create a PR to adjust it?

**Comment by aminvakil** [comment]
@felixfontein Sure, please accept my apologies for bringing it up after this got merged.

**Comment by felixfontein** [comment]
@aminvakil no need to apologize! It's still plenty of time until the 2.0.0 release, and it only matters that it is done before that ;)

**Pull request merged**
